src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,447 @@
+    return current_size > max_digestion_size;
+}
+
+void GinIndexStore::finalize()
+{
+    if (!current_postings.empty())
+    {
+        writeSegment();
+    }
+}
+
+void GinIndexStore::init_file_streams()
+{
+    String segment_file_name = getName() + ".gin_seg";
+    String item_dict_file_name = getName() + ".gin_dict";
+    String postings_file_name = getName() + ".gin_post";
+
+    segment_file_stream = data_part_storage_builder->writeFile(segment_file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append, {});
+    term_dict_file_stream = data_part_storage_builder->writeFile(item_dict_file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append, {});
+    postings_file_stream = data_part_storage_builder->writeFile(postings_file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append, {});
+}
+
+void GinIndexStore::writeSegment()
+{
+    if (segment_file_stream == nullptr)
+    {
+        init_file_streams();
+    }
+
+    ///write segment
+    segment_file_stream->write(reinterpret_cast<char*>(&current_segment), sizeof(GinIndexSegment));
+    std::vector<std::pair<std::string_view, GinIndexPostingsBuilderPtr>> token_postings_list_pairs;
+    token_postings_list_pairs.reserve(current_postings.size());
+
+    for (const auto& [token, postings_list] : current_postings)
+    {
+        token_postings_list_pairs.push_back({std::string_view(token), postings_list});
+    }
+    std::sort(token_postings_list_pairs.begin(), token_postings_list_pairs.end(),
+                    [](const std::pair<std::string_view, GinIndexPostingsBuilderPtr>& a, const std::pair<std::string_view, GinIndexPostingsBuilderPtr>& b)
+                    {
+                        return a.first < b.first;
+                    });
+
+    ///write postings
+    std::vector<UInt64> encoding_lengths(current_postings.size(), 0);
+    size_t current_index = 0;
+
+    for (const auto& [token, postings_list] : token_postings_list_pairs)
+    {
+        auto encoding_length = postings_list->serialize(*postings_file_stream);
+
+        encoding_lengths[current_index++] = encoding_length;
+        current_segment.postings_start_offset += encoding_length;
+    }
+    ///write item dictionary
+    std::vector<UInt8> buffer;
+    WriteBufferFromVector<std::vector<UInt8>> write_buf(buffer);
+    FSTBuilder builder(write_buf);
+
+    UInt64 offset{0};
+    current_index = 0;
+    for (const auto& [token, postings_list] : token_postings_list_pairs)
+    {
+        String str_token{token};
+        builder.add(str_token, offset);
+        offset += encoding_lengths[current_index++];
+    }
+
+    builder.build();
+    write_buf.finalize();
+
+    /// Write FST size
+    writeVarUInt(buffer.size(), *term_dict_file_stream);
+    current_segment.item_dict_start_offset += getLengthOfVarUInt(buffer.size());
+
+    /// Write FST content
+    term_dict_file_stream->write(reinterpret_cast<char*>(buffer.data()), buffer.size());
+    current_segment.item_dict_start_offset += buffer.size();
+
+    current_size = 0;
```

Comment:
If there is an exception anywhere above, the member variables (l. 284-286) are not reset. That's a problem if GinIndexStore::writeSegment() is called again. Maybe put in a try/catch?

Reply:
According to my understanding, if there are exceptions during skipping index building process, the index(which are some temporary files) will be discarded and new temporary index will be created from scratch, so it seems try and catch not necessary IMHO.